README.md
# i0

**i0** (pronounced "i-zero") provides a simple interface for exploring interactions in linear models. This R package is an implementation of the methods in Page-Gould and Miller (submitted), which extends Aiken and West (1991).

## usage

```R
# load the i0 library
library(i0)

# the following dataset comes with the i0 package
data(candy_example)

# create a zero-targeted model using the dataset
zt = target(candy ~ age + height + age * height, data=candy_example)

# explore +/- 1 standard deviation like Aiken and West (1991)
plot(zt)
```

![gaussian results](http://iandennismiller.github.io/i0/gaussian_sim.png)

## installation

The **i0** R package is easily installed using R or RStudio. In the R console, load **devtools** and install **i0** from the project website:

```R
library(devtools)
install_github("i0", "iandennismiller")
```

If you receive the message `Error in library(devtools) : there is no package called devtools`, you must first install **devtools** (see instructions below) then try again.

```R
install.packages("devtools")
```

## documentation

- [Examples](https://github.com/iandennismiller/i0/wiki/Examples)
- [Project Website](http://iandennismiller.github.io/i0)
- [Source Code](https://github.com/iandennismiller/i0)

## support

Please visit the [issue tracker](https://github.com/iandennismiller/i0/issues) on github to report a problem.

## references

Aiken, L. S., & West, S. G. (1991). Multiple regression: Testing and interpreting interactions. Sage Publications, Inc.

Page-Gould, E., & Miller, I. D. (submitted). Zeroing in on the intercept: Estimating marginal means and standard errors with linear models.


NEWS
i0 0.1
===========

* new implementation based on Page-Gould & Miller (2013).


tests/main.R
library('i0')
data(cats, package="MASS")
x = cbind(Const=1, Bwt=cats$Bwt)
y = cats$Hw
mod1 <- i0(x, y)
mod1
summary(mod1)
summary(i0(Hwt~Bwt*Sex, data=cats))


tests/test-all.R
library(testthat)
library(i0)

test_package("i0")


tests/installation.R
library(devtools)
dev_mode()
install_github("iandennismiller/i0")


R/legacy.R
library(ggplot2)

just_summary <- function(spec, distro_family) {
  model <- lmer(spec, na.action="na.exclude", family=distro_family)
  print(summary(model))
}

extract_mean <- function(spec, distro_family) {
  model <- lmer(spec, na.action="na.exclude", family=distro_family)
  ret_mean <- attr(model, 'fixef')[[1]]
  ret_stderr <- attr(summary(model), "coefs")[[1,2]]
  return(list(ret_mean, ret_stderr))
}

extract_stderr <- function(model) {
  return(0)
}

simple_slopes_core <- function (ids, dv, dim1, dim2, distro_family) {
  simple_slopes <- data.frame(ids)
  simple_slopes$dim1_high <- dim1 - sd(dim1, na.rm=T)
  simple_slopes$dim1_low <- dim1 + sd(dim1, na.rm=T)
  simple_slopes$dim2_high <- dim2 - sd(dim2, na.rm=T)
  simple_slopes$dim2_low <- dim2 + sd(dim2, na.rm=T)
  return(simple_slopes)
}

# Use Page-Gould's quick means technique to find the 4 means
simple_slopes_means <- function (ids, dv, distro_family, ss) {
  means <- list()
  means$low_low <- extract_mean(dv ~ (1|ids) + ss$dim1_low * ss$dim2_low, distro_family)[[1]]
  means$low_high <- extract_mean(dv ~ (1|ids) + ss$dim1_low * ss$dim2_high, distro_family)[[1]]
  means$high_low <- extract_mean(dv ~ (1|ids) + ss$dim1_high * ss$dim2_low, distro_family)[[1]]
  means$high_high <- extract_mean(dv ~ (1|ids) + ss$dim1_high * ss$dim2_high, distro_family)[[1]]

  stderr <- list()
  stderr$low_low <- extract_mean(dv ~ (1|ids) + ss$dim1_low * ss$dim2_low, distro_family)[[2]]
  stderr$low_high <- extract_mean(dv ~ (1|ids) + ss$dim1_low * ss$dim2_high, distro_family)[[2]]
  stderr$high_low <- extract_mean(dv ~ (1|ids) + ss$dim1_high * ss$dim2_low, distro_family)[[2]]
  stderr$high_high <- extract_mean(dv ~ (1|ids) + ss$dim1_high * ss$dim2_high, distro_family)[[2]]

  return(list(means, stderr))
}

# make sure both dim1 and dim2 have been scaled
# e.g. simple_slopes(scale(closeness, scale=F), scale(predictions, scale=F))
simple_slopes_tests <- function (ids, dv, dim1, dim2, distro_family) {
  ss <- simple_slopes_core(ids, dv, dim1, dim2, distro_family)
  cat("\n########### dim1 high\n\n")
  just_summary(dv ~ (1|ids) + ss$dim1_high * dim2, distro_family)
  cat("\n########### dim1 low\n\n")
  just_summary(dv ~ (1|ids) + ss$dim1_low * dim2, distro_family)
  cat("\n########### dim2 high\n\n")
  just_summary(dv ~ (1|ids) + ss$dim2_high * dim1, distro_family)
  cat("\n########### dim2 low\n\n")
  just_summary(dv ~ (1|ids) + ss$dim2_low * dim1, distro_family)
}

simple_slopes_plot <- function(ids, dv, dim1, dim2, distro_family, dv_name, dim1_name, dim2_name) {
  ss <- simple_slopes_core(ids, dv, dim1, dim2, distro_family)
  means <- simple_slopes_means(ids, dv, distro_family, ss)[[1]]
  stderr <- simple_slopes_means(ids, dv, distro_family, ss)[[2]]

  display <- data.frame(
    dim1 = c('low', 'low', 'high', 'high'),
    dim2 = c('low', 'high', 'low', 'high'),
    means = c(means$low_low, means$low_high, means$high_low, means$high_high),
    ci = c(stderr$low_low, stderr$low_high, stderr$high_low, stderr$high_high)
  )
  display$dim2 <- factor(display$dim2, levels=c('low', 'high'))

  ggplot(display, aes(x=dim2, y=means, group=dim1)) +
    geom_line(aes(linetype=dim1), size=1) +
    geom_point(size=3, fill="white") +
    scale_x_discrete(labels=c(paste("low", dim2_name), paste("high", dim2_name))) +
    scale_linetype(labels=c(paste("high", dim1_name), paste("low", dim1_name))) +
    coord_cartesian(xlim=c(0.8, 2.2)) +
    xlab(NULL)+
    ylab(dv_name) +
    geom_errorbar(aes(ymin=means-ci, ymax=means+ci), alpha=0.3, width=.05) +
    #theme_bw() +
    opts(title = paste(dv_name, " ~ '", dim1_name, "' and '", dim2_name, "'", sep="")) +
    opts(legend.title=theme_blank()) + opts(legend.position=c(.2, .9)) +
    opts(legend.text = theme_text(size = 12)) +
    opts(axis.text.x = theme_text(size = 14)) +
    opts(axis.text.y = theme_text(size = 14))
}

simple_slopes <- function(ids, dv, dim1, dim2, distro_family, dv_name, dim1_name, dim2_name) {
  #simple_slopes_tests(ids, dv, dim1, dim2, distro_family)
  simple_slopes_plot(ids, dv, dim1, dim2, distro_family, dv_name, dim1_name, dim2_name)
}

simple_slopes_3way = function(dv_name, dim1_name, dim2_name, dim3_name, data, family="gaussian") {
  with(data, {
    dv = get(dv_name)
    dim1 = get(dim1_name)
    dim2 = get(dim2_name)
    dim3 = get(dim3_name)

    display = data.frame(dim1 = NA, dim2 = NA, dim3 = NA, means = NA, ci = NA)

    dims = list()
    dims[[1]] = dims[[2]] = dims[[3]] = list()

    dims[[1]][["high"]] = dim1 - sd(dim1, na.rm=T)
    dims[[1]][["low"]] = dim1 + sd(dim1, na.rm=T)
    dims[[2]][["high"]] = dim2 - sd(dim2, na.rm=T)
    dims[[2]][["low"]] = dim2 + sd(dim2, na.rm=T)
    dims[[3]][["high"]] = dim3 - sd(dim3, na.rm=T)
    dims[[3]][["low"]] = dim3 + sd(dim3, na.rm=T)

    spec_dim1_high = dv ~ dim3 * dim2 * dims[[1]][["high"]]
    spec_dim1_low = dv ~ dim3 * dim2 * dims[[1]][["low"]]
    summary_dim1_high = summary(glm(spec_dim1_high, family=family))
    summary_dim1_low = summary(glm(spec_dim1_low, family=family))

    for (level in c("low", "high")) {
      spec = dv ~ dim3 * dim2 * dims[[1]][[level]]
      summary_coefficients = summary(glm(spec, family=family))$coefficients
      if (summary_coefficients[5,4] < 0.05) {
        cat(paste("with", dim1_name, level, "p=", summary_coefficients[5,4], "\n"))

        for (sublevel in c("low", "high")) {
          spec_sublevel = dv ~ dims[[3]][[sublevel]] * dim2 * dims[[1]][[level]]
          summary_sublevel_coefficients = summary(glm(spec_sublevel, family=family))$coefficients

          if (summary_sublevel_coefficients[[3, 4]] < 0.05) {
            cat(paste("with", dim1_name, level, "and", dim3_name, sublevel, "p=",
                    summary_sublevel_coefficients[[3, 4]], "\n"))
          }
        }

        spec = intercepts = stderr = list()

        spec[[1]] = dv ~ dims[[3]][["high"]] * dims[[2]][["low"]] * dims[[1]][[level]]
        spec[[2]] = dv ~ dims[[3]][["low"]] * dims[[2]][["low"]] * dims[[1]][[level]]
        spec[[3]] = dv ~ dims[[3]][["high"]] * dims[[2]][["high"]] * dims[[1]][[level]]
        spec[[4]] = dv ~ dims[[3]][["low"]] * dims[[2]][["high"]] * dims[[1]][[level]]

        for (i in 1:4) {
          fit = summary(glm(spec[[i]], family=family))
          intercepts[[i]] = fit$coefficients[1,1]
          stderr[[i]] = fit$coefficients[1,2]
        }

        display <- merge(display, data.frame(
          dim1 = rep(level, 4),
          dim2 = c('high', 'high', 'low', 'low'),
          dim3 = c('high', 'low', 'high', 'low'),
          means = as.numeric(intercepts),
          ci = as.numeric(stderr)
        ), all=T)
      }
    }

    display$dim1 <- factor(display$dim1, levels=c('low', 'high'))
    display$dim2 <- factor(display$dim2, levels=c('low', 'high'))
    display$dim3 <- factor(display$dim3, levels=c('low', 'high'))
    display = display[1:8,]

    ggplot(display, aes(x=dim3, y=means, group=dim2)) +
      facet_grid(. ~ dim1) +
      geom_line(aes(linetype=dim2), size=1) +
      geom_point(size=3, fill="white") +
      scale_x_discrete(labels=c("low", "high")) +
      scale_linetype(labels=c(paste("high\n", dim2_name), paste("low\n", dim2_name))) +
      coord_cartesian(xlim=c(0.8, 2.2)) +
      xlab(dim3_name) +
      ylab(dv_name) +
      geom_errorbar(aes(ymin=means-ci, ymax=means+ci), alpha=0.3, width=.05) +
      opts(title = paste(dv_name, "~", dim1_name, "*\n", dim2_name, "*", dim3_name, sep=" ")) +
      opts(plot.title = theme_text(size=10)) +
      opts(legend.title=theme_blank()) +
      opts(legend.position="bottom") +
      opts(legend.text = theme_text(size = 11)) +
      opts(axis.text.x = theme_text(size = 14)) +
      opts(axis.text.y = theme_text(size = 14))
  })
}


R/target.R
#' Create a target object.
#'
#' @param x input model function specification
#' @return target object
#' @keywords character
#' @export
target <- function(x, ...) UseMethod("target")

######
# generic handlers

#' Print a zero-targeted model
#' 
#' @param x a zero-targeted model
#' @export
print.target <- function(x, ...)
{
    print(x$estimates)
}

#' Get mean and stderr information about a zero-targeted model
#' 
#' @param object a zero-targeted model
#' @export
summary.target <- function(object, ...)
{
    res = list(summary = summarize(object))
    class(res) <- "summary.target"
res
}

#' Print a zero-targeted summary
#'
#' @param x a zero-targeted summary object
#' @export
print.summary.target <- function(x, ...)
{
    print(x$summary)
}

#' Plot a zero-targeted model +/- 1 standard deviation.
#' 
#' @param x a zero-targeted model object
#' @param raw boolean indicating whether to plot raw values; defaults to FALSE 
#'  (i.e. print scaled rather than raw values)
#' @export
plot.target <- function(x, raw=F, ...) {
    display = summarize(x)
    if (raw) {
        display = transform_raw(display, x$family)
    }
    else {
        display$ymins = display$means - display$ci
        display$ymaxs = display$means + display$ci
    }

    ggplot(display, aes(x=dim2, y=means, group=dim1)) +
        geom_line(aes(linetype=dim1), size=1) +
        geom_point(size=3, fill="white") +
        scale_x_discrete(labels=c(paste("low", x$terms$d2_name), paste("high", x$terms$d2_name))) +
        scale_linetype(labels=c(paste("high", x$terms$d1_name), paste("low", x$terms$d1_name))) +
        coord_cartesian(xlim=c(0.8, 2.2)) +
        xlab(NULL)+
        ylab(x$terms$dv_name) +
        theme_bw() +
        geom_errorbar(aes(ymin=ymins, ymax=ymaxs), alpha=0.3, width=.05) +
        labs(title = paste(x$terms$dv_name, " ~ '", x$terms$d1_name, "' and '", x$terms$d2_name, "'", sep="")) +
        theme(legend.title=element_blank()) +
        theme(legend.position=c(.2, .9)) +
        theme(legend.text = element_text(size = 12)) +
        theme(axis.text.x = element_text(size = 14)) +
        theme(axis.text.y = element_text(size = 14))
}

#################
# private methods

# Unpack an S3 formula to figure out the name of the DV and IVs
#
# @param formula model specification
# @return target object
unpack_formula <- function(formula) {
    terms = names(attr(terms(formula), "factors")[,1])
    list(
        dv_name = terms[1],
        d1_name = terms[2],
        d2_name = terms[3]
    )
}

# undocumented
#
transform_raw <- function(display, family) {
    if (family == "poisson") {
        b0 = display$means - display$ci
        display$ymins = exp(b0)
        b0 = display$means + display$ci
        display$ymaxs = exp(b0)
        b0 = display$means
        display$means = exp(b0)
    }
    else if (family == "binomial") {
        b0 = display$means - display$ci
        display$ymins = exp(b0) / (1 + exp(b0))
        b0 = display$means + display$ci
        display$ymaxs = exp(b0) / (1 + exp(b0))
        b0 = display$means
        display$means = exp(b0) / (1 + exp(b0))
    }
    else {
        display$ymins = display$means - display$ci
        display$ymaxs = display$means + display$ci
    }
    display
}

# undocumented
#
summarize <- function(x, ...) {
    display = data.frame(
        dim1 = c('low', 'low', 'high', 'high'),
        dim2 = c('low', 'high', 'low', 'high'),
        means = c(
            x$estimates$low$low$mean,
            x$estimates$low$high$mean,
            x$estimates$high$low$mean,
            x$estimates$high$high$mean
        ),
        ci = c(
            x$estimates$low$low$stderr,
            x$estimates$low$high$stderr,
            x$estimates$high$low$stderr,
            x$estimates$high$high$stderr
        )
    )

    display$dim2 = factor(display$dim2, levels=c('low', 'high'))
    display
}

# undocumented
#
calc_zero_target <- function(terms, data) {
    # zt means "zero targeted"
    zt = data.frame(
        #ids = seq(from=1, to=dim(data)[1]),
        dv = data[[terms$dv_name]],
        d1_high = data[[terms$d1_name]] - sd(data[[terms$d1_name]], na.rm=T),
        d1_low = data[[terms$d1_name]] + sd(data[[terms$d1_name]], na.rm=T),
        d2_high = data[[terms$d2_name]] - sd(data[[terms$d2_name]], na.rm=T),
        d2_low = data[[terms$d2_name]] + sd(data[[terms$d2_name]], na.rm=T)
        )
    data.frame(data, zt)
}

# undocumented
#
gen_formulas <- function(formula_str, terms) {
    f_low_low = str_replace_all(formula_str, terms$d1_name, "d1_low")
    f_low_low = str_replace_all(f_low_low, terms$d2_name, "d2_low")

    f_low_high = str_replace_all(formula_str, terms$d1_name, "d1_low")
    f_low_high = str_replace_all(f_low_high, terms$d2_name, "d2_high")

    f_high_low = str_replace_all(formula_str, terms$d1_name, "d1_high")
    f_high_low = str_replace_all(f_high_low, terms$d2_name, "d2_low")

    f_high_high = str_replace_all(formula_str, terms$d1_name, "d1_high")
    f_high_high = str_replace_all(f_high_high, terms$d2_name, "d2_high")

    list(
        low_low = as.formula(f_low_low),
        low_high = as.formula(f_low_high),
        high_low = as.formula(f_high_low),
        high_high = as.formula(f_high_high)
    )
}

# undocumented
#
calc_estimates <- function(f, zt, mlm, family) {
    if (mlm) {
        fn = calc_lmer
    }
    else {
        fn = calc_lm
    }

    list(
        low_low = fn(f$low_low, zt, family),
        low_high = fn(f$low_high, zt, family),
        high_low = fn(f$high_low, zt, family),
        high_high = fn(f$high_high, zt, family)
    )
}

# undocumented
#
unpack_estimates <- function(estimate) {
    list(
        low = list(
            low = list(
                mean = estimate[["low_low"]][["mean"]],
                stderr = estimate[["low_low"]][["stderr"]]
                ),
            high = list(
                mean = estimate[["low_high"]][["mean"]],
                stderr = estimate[["low_high"]][["stderr"]]
                )
            ),
        high = list(
            low = list(
                mean = estimate[["high_low"]][["mean"]],
                stderr = estimate[["high_low"]][["stderr"]]
                ),
            high = list(
                mean = estimate[["high_high"]][["mean"]],
                stderr = estimate[["high_high"]][["stderr"]]
                )
            )
        )
}

# undocumented
#
calc_lm <- function(spec, data, family) {
    model = glm(spec, data, na.action="na.exclude", family=family)
    list(
        model = model,
        mean = summary(model)$coefficients[1,1],
        stderr = summary(model)$coefficients[1,2]
    )
}

# undocumented
#
calc_lmer <- function(spec, data, family) {
    model = lmer(spec, data, na.action="na.exclude", family=family)
    list(
        model = model,
        mean = attr(model, 'fixef')[[1]],
        stderr = attr(summary(model), "coefs")[[1,2]]
    )
}

################
# public methods

#' Create a target object.
#'
#' @param x input model function specification
#' y data frame
#' @return target object
#' @keywords character
#' @export
#' @examples
#' data = data.frame(y = c(0, 1, 2), x1 = c(2, 4, 6), x2 = c(3, 6, 9))
#' t = target(y ~ x1 * x2, data=data)
#' t$point
target.formula <- function(
    formula,
    data=list(),
    family="gaussian",
    mlm=FALSE,
    ...)
{
    terms = unpack_formula(formula)
    f_str = as.character(formula)
    formula_str = paste(f_str[2], f_str[1], f_str[3])
    std_formulas = gen_formulas(formula_str, terms)
    zt = calc_zero_target(terms, data)
    models = calc_estimates(std_formulas, zt, mlm, family)
    obj = list(
        formula = formula,
        call = match.call(),
        terms = terms,
        zero_targeted = zt,
        models = models,
        family = family,
        estimates = unpack_estimates(models)
        )
    class(obj) = "target"
    obj
}
